File: packages/nuclear_data_to_arrow/src/nuclear_data_to_arrow/synthesis.py
"""Synthesize hierarchical MTs and build FastXSGrid lookup tables.

This module implements yamc's post-processing algorithms in Python so that
the Arrow files contain simulation-ready data.
"""

import numpy as np

from endf.data import SUM_RULES


# MT numbers that count as scattering (non-inelastic subset).
#
# This is yamc's own classification for the per-channel scatter breakdown in
# fast_xs, not an ENDF summation rule, so it is defined here rather than taken
# from SUM_RULES. Every channel that re-emits a neutron belongs in it.
#
# The 152-200 range is a convenient way to reach the multi-particle channels, but
# it also sweeps up the few in that range that emit no neutron at all: (n,ta),
# (n,dt), (n,p3He), (n,d3He), (n,3Hea) and (n,3p). Those are disappearance
# reactions and belong to MT 101 instead. Since MT 3 is built as
# scattering + fission + MT 101, leaving them in both groups counts them twice,
# so they are subtracted out. See issue #23.
_SCATTERING_CANDIDATES = frozenset([
    2, 5, 11, 16, 17, 22, 23, 24, 25, 28, 29, 30,
    32, 33, 34, 35, 36, 37, 41, 42, 44, 45,
    *range(152, 201),
    *range(875, 891),
])
SCATTERING_MTS_NON_INELASTIC = _SCATTERING_CANDIDATES - frozenset(SUM_RULES[101])

# Inelastic scattering: MT 50-91
INELASTIC_MTS = frozenset(SUM_RULES[4])

# All scattering MTs
ALL_SCATTERING_MTS = INELASTIC_MTS | SCATTERING_MTS_NON_INELASTIC

# Fission MTs: the total (18) and its partials
FISSION_MTS = frozenset([18, *SUM_RULES[18]])

# MTs that contribute to MT 101, neutron disappearance.
#
# Taken from the ENDF summation rules rather than spelled out here. A private
# copy previously included 13 channels that re-emit a neutron, such as (n,npd),
# (n,2n2p) and (n,n3p), which do not belong in a disappearance sum. See #23.
ABSORPTION_MTS = frozenset(SUM_RULES[101])

# MT 3 is built as scattering + fission + MT 101, so the three groups have to be
# disjoint or a channel is counted twice. Asserted here so any future change to
# the sets fails loudly rather than quietly inflating MT 3 and MT 1.
assert not (ABSORPTION_MTS & (INELASTIC_MTS | SCATTERING_MTS_NON_INELASTIC))
assert not (ABSORPTION_MTS & FISSION_MTS)
assert not ((INELASTIC_MTS | SCATTERING_MTS_NON_INELASTIC) & FISSION_MTS)

# Synthesized MT numbers -- these are redundant sums
SYNTHETIC_MTS = frozenset([1, 3, 4, 27, 101])

# Number of log bins for FastXSGrid
N_LOG_BINS = 8000


def is_scattering_mt(mt):
    """Return True if the given MT is a scattering reaction."""
    return mt in ALL_SCATTERING_MTS


def is_fission_mt(mt):
    """Return True if the given MT is a fission reaction."""
    return mt in FISSION_MTS


def _interp_xs_to_grid(reaction, temperature, energy_grid):
    """Interpolate a reaction's cross section onto the full energy grid.

    Parameters
    ----------
    reaction : endf.Reaction
        The reaction object.
    temperature : str
        Temperature key (e.g., "294K").
    energy_grid : numpy.ndarray
        The full energy grid for this temperature.

    Returns
    -------
    numpy.ndarray
        Cross-section values on the full energy grid.
    """
    xs_tab = reaction.xs.get(temperature)
    if xs_tab is None:
        return np.zeros(len(energy_grid), dtype=np.float64)

    # xs_tab is a Tabulated1D-like with .x, .y and a _threshold_idx
    threshold_idx = getattr(xs_tab, '_threshold_idx', 0)
    xs_y = np.asarray(xs_tab.y, dtype=np.float64)

    # Build result array (zeros below threshold)
    result = np.zeros(len(energy_grid), dtype=np.float64)

    if len(xs_y) == 0:
        return result

    # The xs_y array starts at threshold_idx into the energy grid
    n_xs = len(xs_y)
    end_idx = threshold_idx + n_xs
    if end_idx <= len(energy_grid):
        result[threshold_idx:end_idx] = xs_y
    else:
        # Clip to available grid length
        avail = len(energy_grid) - threshold_idx
        result[threshold_idx:] = xs_y[:avail]

    return result


def synthesize_hierarchical_mts(data, temperature):
    """Synthesize hierarchical MT cross sections from constituent reactions.

    Parameters
    ----------
    data : endf.IncidentNeutron
        The incident neutron data object.
    temperature : str
        Temperature key (e.g., "294K").

    Returns
    -------
    dict
        Mapping of MT number to numpy cross-section array on the full energy
        grid.  Keys are 4, 101, 27, 3, 1.  MT 1 and 101 are always present
        (even if all zeros).
    """
    energy_grid = np.asarray(data.energy[temperature], dtype=np.float64)
    n_energy = len(energy_grid)

    # Pre-interpolate all reactions onto the full grid
    rx_xs = {}
    for mt, rx in data.reactions.items():
        if mt not in SYNTHETIC_MTS:
            rx_xs[mt] = _interp_xs_to_grid(rx, temperature, energy_grid)

    result = {}

    # MT 4 = sum(MT 50-91) -- total inelastic
    mt4 = np.zeros(n_energy, dtype=np.float64)
    for mt in INELASTIC_MTS:
        if mt in rx_xs:
            mt4 += rx_xs[mt]
    result[4] = mt4

    # MT 101 = sum(MT 102-117, 155, 182-199) -- absorption excluding fission
    mt101 = np.zeros(n_energy, dtype=np.float64)
    for mt in ABSORPTION_MTS:
        if mt in rx_xs:
            mt101 += rx_xs[mt]
    result[101] = mt101

    # MT 18 total fission (use if present, otherwise sum partial fission)
    fission_xs = np.zeros(n_energy, dtype=np.float64)
    if 18 in rx_xs:
        fission_xs = rx_xs[18].copy()
    else:
        for mt in [19, 20, 21, 38]:
            if mt in rx_xs:
                fission_xs += rx_xs[mt]

    # MT 27 = MT 18 + MT 101 -- total absorption
    result[27] = fission_xs + mt101

    # MT 3 = non-elastic (all scattering except elastic + fission + absorption)
    # = sum of all non-synthetic, non-redundant reactions that are scattering
    #   + fission + MT 101
    mt3 = np.zeros(n_energy, dtype=np.float64)
    for mt, xs in rx_xs.items():
        if mt in SYNTHETIC_MTS:
            continue
        if is_scattering_mt(mt) and mt != 2:
            mt3 += xs
    mt3 += fission_xs + mt101
    result[3] = mt3

    # MT 1 = MT 2 (elastic) + MT 3 (non-elastic) -- total
    elastic = rx_xs.get(2, np.zeros(n_energy, dtype=np.float64))
    result[1] = elastic + mt3

    return result


def build_fast_xs(data, temperature, synthetic_mts):
    """Build FastXSGrid arrays for fast cross-section lookup.

    Parameters
    ----------
    data : endf.IncidentNeutron
        The incident neutron data object.
    temperature : str
        Temperature key (e.g., "294K").
    synthetic_mts : dict
        Mapping from MT number to numpy XS arrays (from
        ``synthesize_hierarchical_mts``).

    Returns
    -------
    dict
        Dictionary with FastXSGrid data:
        - ``log_e_min`` (float): log of minimum energy
        - ``inv_log_delta`` (float): inverse of log bin width
        - ``log_grid_index`` (ndarray): grid index per log bin (N_LOG_BINS+1)
        - ``xs`` (ndarray): shape (n_energy, 4) -- [total, absorption, scatter, fission]
        - ``energy`` (ndarray): the full energy grid
        - ``scatter_mt_numbers`` (list of int): MT numbers for scattering channels
        - ``scatter_mt_xs`` (ndarray): shape (n_energy, n_scatter) -- XS per scatter MT
        - ``fission_mt_numbers`` (list of int): MT numbers for fission channels
        - ``fission_mt_xs`` (ndarray): shape (n_energy, n_fission) -- XS per fission MT
        - ``has_partial_fission`` (bool): True if partial fission MTs present
        - ``xs_ngamma`` (ndarray): (n,gamma) capture XS (MT 102)
        - ``photon_prod`` (ndarray): photon production XS
    """
    energy_grid = np.asarray(data.energy[temperature], dtype=np.float64)
    n_energy = len(energy_grid)

    # Build the 4-column XS array: [total, absorption, scatter, fission]
    mt1 = synthetic_mts.get(1, np.zeros(n_energy, dtype=np.float64))
    mt101 = synthetic_mts.get(101, np.zeros(n_energy, dtype=np.float64))

    # Fission XS
    fission_xs = np.zeros(n_energy, dtype=np.float64)
    rx_xs_cache = {}
    for mt, rx in data.reactions.items():
        if mt not in SYNTHETIC_MTS:
            rx_xs_cache[mt] = _interp_xs_to_grid(rx, temperature, energy_grid)

    if 18 in rx_xs_cache:
        fission_xs = rx_xs_cache[18].copy()
    else:
        for mt in [19, 20, 21, 38]:
            if mt in rx_xs_cache:
                fission_xs += rx_xs_cache[mt]

    # Absorption = MT 101 (disappearance, excluding fission)
    # Fission is stored as a separate column
    absorption = mt101.copy()

    # Scattering = total - absorption - fission
    scattering = mt1 - absorption - fission_xs
    # Clamp to zero (numerical noise)
    scattering = np.maximum(scattering, 0.0)

    xs_4col = np.column_stack([mt1, absorption, scattering, fission_xs])

    # Build log grid index
    log_e_min = np.log(energy_grid[0]) if energy_grid[0] > 0 else -100.0
    log_e_max = np.log(energy_grid[-1]) if energy_grid[-1] > 0 else 100.0
    log_delta = (log_e_max - log_e_min) / N_LOG_BINS
    inv_log_delta = 1.0 / log_delta if log_delta > 0 else 0.0

    log_grid_index = np.zeros(N_LOG_BINS + 1, dtype=np.int32)
    log_energy = np.log(energy_grid)

    for i in range(N_LOG_BINS + 1):
        log_e = log_e_min + i * log_delta
        # Find the index in energy_grid where log(E) >= log_e
        idx = int(np.searchsorted(log_energy, log_e, side='right')) - 1
        log_grid_index[i] = max(0, min(idx, n_energy - 1))

    # Scatter MT breakdown
    scatter_mt_numbers = []
    scatter_mt_xs_list = []
    for mt in sorted(rx_xs_cache.keys()):
        if is_scattering_mt(mt):
            scatter_mt_numbers.append(mt)
            scatter_mt_xs_list.append(rx_xs_cache[mt])

    if scatter_mt_xs_list:
        scatter_mt_xs = np.column_stack(scatter_mt_xs_list)
    else:
        scatter_mt_xs = np.empty((n_energy, 0), dtype=np.float64)

    # Fission MT breakdown
    fission_mt_numbers = []
    fission_mt_xs_list = []
    has_partial_fission = False
    for mt in [19, 20, 21, 38]:
        if mt in rx_xs_cache:
            fission_mt_numbers.append(mt)
            fission_mt_xs_list.append(rx_xs_cache[mt])
            has_partial_fission = True

    if not has_partial_fission and 18 in rx_xs_cache:
        fission_mt_numbers = [18]
        fission_mt_xs_list = [rx_xs_cache[18]]

    if fission_mt_xs_list:
        fission_mt_xs = np.column_stack(fission_mt_xs_list)
    else:
        fission_mt_xs = np.empty((n_energy, 0), dtype=np.float64)

    # (n,gamma) capture -- MT 102
    xs_ngamma = rx_xs_cache.get(102, np.zeros(n_energy, dtype=np.float64))

    # Photon production XS -- sum of all photon-producing reaction XS
    # For simplicity, use zeros; yamc computes this from product data
    photon_prod = np.zeros(n_energy, dtype=np.float64)

    return {
        "log_e_min": float(log_e_min),
        "inv_log_delta": float(inv_log_delta),
        "log_grid_index": log_grid_index,
        "xs": xs_4col,
        "energy": energy_grid,
        "scatter_mt_numbers": scatter_mt_numbers,
        "scatter_mt_xs": scatter_mt_xs,
        "fission_mt_numbers": fission_mt_numbers,
        "fission_mt_xs": fission_mt_xs,
        "has_partial_fission": has_partial_fission,
        "xs_ngamma": xs_ngamma,
        "photon_prod": photon_prod,
    }

File: crates/yamc-convert/src/synthesis.rs
//! The redundant MTs, built from their partials.
//!
//! An evaluation gives the partial channels; a transport code wants the sums.
//! MT 1 (total), 3 (non-elastic), 4 (inelastic), 27 (absorption) and 101
//! (disappearance) are therefore synthesized at conversion time rather than at
//! every simulation start.
//!
//! The sets below are taken from the ENDF summation rules in [`endf::data`]
//! rather than written out here. That is not tidiness: a hand-written copy of
//! the MT 101 set once included thirteen channels that re-emit a neutron, such
//! as (n,npd), (n,2n2p) and (n,n3p), which do not belong in a disappearance
//! sum. They were counted into MT 101, and through it into MT 27, MT 3 and
//! MT 1, so a nuclide's total came out above its own evaluated total. Deriving
//! the sets from one source means a change to the rules cannot leave two
//! definitions disagreeing.

use std::collections::{BTreeMap, BTreeSet};

use endf::data::sum_rule;

/// The MTs this module builds. Excluded from every sum, since summing a sum
/// double counts.
pub const SYNTHETIC_MTS: [i32; 5] = [1, 3, 4, 27, 101];

/// Channels that emit a neutron and are not level inelastic.
///
/// The candidates, less anything the disappearance rule claims: (n,2n) and
/// friends belong here, and the (n,npd)-shaped channels belong in exactly one
/// of the two groups, never both.
fn scattering_non_inelastic() -> BTreeSet<i32> {
    let mut set: BTreeSet<i32> = [
        2, 5, 11, 16, 17, 22, 23, 24, 25, 28, 29, 30, 32, 33, 34, 35, 36, 37, 41, 42, 44, 45,
    ]
    .into_iter()
    .chain(152..201)
    .chain(875..891)
    .collect();
    for mt in absorption() {
        set.remove(&mt);
    }
    set
}

/// Level inelastic scattering, MT 50 to 91.
fn inelastic() -> BTreeSet<i32> {
    sum_rule(4).unwrap_or(&[]).iter().copied().collect()
}

/// Neutron disappearance: capture and charged-particle-out channels.
fn absorption() -> BTreeSet<i32> {
    sum_rule(101).unwrap_or(&[]).iter().copied().collect()
}

/// The fission total and its partials.
fn fission() -> BTreeSet<i32> {
    let mut set: BTreeSet<i32> = sum_rule(18).unwrap_or(&[]).iter().copied().collect();
    set.insert(18);
    set
}

/// Whether an MT emits a neutron, either as level inelastic or otherwise.
pub fn is_scattering(mt: i32) -> bool {
    inelastic().contains(&mt) || scattering_non_inelastic().contains(&mt)
}

/// The three groups MT 3 is built from must be disjoint, or a channel is
/// counted twice and every sum above it is wrong.
///
/// Checked at run time rather than left as a comment, because the failure is
/// silent: the numbers stay plausible and only disagree with the evaluation's
/// own total by the size of the double-counted channels.
pub fn groups_are_disjoint() -> Result<(), String> {
    let (inel, scat, abs, fiss) = (
        inelastic(),
        scattering_non_inelastic(),
        absorption(),
        fission(),
    );
    let overlap =
        |a: &BTreeSet<i32>, b: &BTreeSet<i32>| -> Vec<i32> { a.intersection(b).copied().collect() };
    for (what, mts) in [
        ("disappearance and scattering", overlap(&abs, &scat)),
        ("disappearance and inelastic", overlap(&abs, &inel)),
        ("disappearance and fission", overlap(&abs, &fiss)),
        ("scattering and fission", overlap(&scat, &fiss)),
        ("inelastic and fission", overlap(&inel, &fiss)),
    ] {
        if !mts.is_empty() {
            return Err(format!(
                "the {what} sets share {mts:?}; MT 3 would double count them"
            ));
        }
    }
    Ok(())
}

/// Place a reaction's cross section on the nuclide's full energy grid.
///
/// A threshold reaction's array starts partway up the grid, at the index the
/// parser recorded, and is zero below it. Getting this offset wrong shifts a
/// whole cross section in energy, which no schema check would notice.
pub fn on_grid(values: &[f64], threshold_idx: usize, n_energy: usize) -> Vec<f64> {
    let mut out = vec![0.0; n_energy];
    if values.is_empty() || threshold_idx >= n_energy {
        return out;
    }
    let n = values.len().min(n_energy - threshold_idx);
    out[threshold_idx..threshold_idx + n].copy_from_slice(&values[..n]);
    out
}

/// Build the redundant MTs from the partials already on the grid.
///
/// `partials` maps MT to a cross section on the full energy grid, and must
/// exclude [`SYNTHETIC_MTS`]: an evaluation carries its own MT 1 and often its
/// own MT 101, and summing those back in would double count everything under
/// them.
/// Whether this MT is a total whose own residual levels are also present.
///
/// MT 16 and MT 875 to 890 both describe (n,2n): the first is the total, the
/// rest are the levels of the residual nucleus. An evaluation may carry both,
/// and adding both counts the reaction twice.
///
/// JEFF-4.0's Be9 is the case that found this. Its evaluation has no MF=3
/// MT 16 at all, so the parser builds MT 16 as the sum of MT 875 to 890 and
/// flags it redundant; summing the flat scattering set then added the sum AND
/// its parts, putting MT 3 (and so the total) 35% above the evaluation's own
/// MT 1 at 7.5 MeV. A total cross section larger than the evaluated total is
/// impossible, which is what makes this a defect rather than a convention.
///
/// One nuclide in 240 published files trips it, which is why a diff against
/// published data could not find it: the Python converter does the same thing,
/// so both were wrong together. OpenMC is immune because it skips redundant
/// reactions before summing (`src/nuclide.cpp`, `if (rx->redundant_) continue`).
fn covered_by_its_own_levels(mt: i32, partials: &BTreeMap<i32, Vec<f64>>) -> bool {
    match sum_rule(mt) {
        Some(rule) => rule.iter().any(|c| partials.contains_key(c)),
        None => false,
    }
}

/// The neutron-emitting channels other than elastic, summed.
///
/// MT 3 is this plus fission plus disappearance, and the `scattering` column of
/// fast_xs.arrow is this plus elastic. Both callers take the sum from here so a
/// change to the rules cannot leave the two disagreeing, the same reason the
/// sets above come from the ENDF rules rather than from a hand-written copy.
///
/// Exposed rather than inlined into [`synthesize`] because the difference is
/// not only tidiness: fast_xs.arrow used to reach the same quantity as
/// `MT 3 - MT 27`, which cancels the absorption back out and loses the low bits
/// of everything smaller than it. See `fast_xs::write_fast_xs`.
pub fn non_elastic_scattering(partials: &BTreeMap<i32, Vec<f64>>, n_energy: usize) -> Vec<f64> {
    let mut out = vec![0.0; n_energy];
    for (&mt, xs) in partials {
        if SYNTHETIC_MTS.contains(&mt) {
            continue;
        }
        // Skip a total whose own levels are in the same sum: taking both
        // counts the channel twice. Same rule the fission branch below uses.
        if mt != 2 && is_scattering(mt) && !covered_by_its_own_levels(mt, partials) {
            for (o, v) in out.iter_mut().zip(xs) {
                *o += v;
            }
        }
    }
    out
}

pub fn synthesize(partials: &BTreeMap<i32, Vec<f64>>, n_energy: usize) -> BTreeMap<i32, Vec<f64>> {
    let zeros = || vec![0.0; n_energy];
    let add = |acc: &mut Vec<f64>, xs: &[f64]| {
        for (a, b) in acc.iter_mut().zip(xs) {
            *a += b;
        }
    };

    let mut out = BTreeMap::new();

    let mut mt4 = zeros();
    for mt in inelastic() {
        if let Some(xs) = partials.get(&mt) {
            add(&mut mt4, xs);
        }
    }

    let mut mt101 = zeros();
    for mt in absorption() {
        if let Some(xs) = partials.get(&mt) {
            add(&mut mt101, xs);
        }
    }

    // MT 18 where the evaluation gives it, otherwise its partials. Never both,
    // which is why this is not a plain sum over the fission set.
    let mut fission_xs = zeros();
    match partials.get(&18) {
        Some(xs) => fission_xs.copy_from_slice(xs),
        None => {
            for mt in [19, 20, 21, 38] {
                if let Some(xs) = partials.get(&mt) {
                    add(&mut fission_xs, xs);
                }
            }
        }
    }

    let mut mt27 = fission_xs.clone();
    add(&mut mt27, &mt101);

    // Non-elastic: every neutron-emitting channel except elastic, plus fission,
    // plus disappearance. Built from the partials present rather than from the
    // set, so a channel the evaluation does not carry contributes nothing.
    let mut mt3 = non_elastic_scattering(partials, n_energy);
    add(&mut mt3, &fission_xs);
    add(&mut mt3, &mt101);

    let mut mt1 = partials.get(&2).cloned().unwrap_or_else(zeros);
    add(&mut mt1, &mt3);

    out.insert(1, mt1);
    out.insert(3, mt3);
    out.insert(4, mt4);
    out.insert(27, mt27);
    out.insert(101, mt101);
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The sets MT 3 is built from must not overlap.
    ///
    /// This is the check that would have caught issue #23 the day it was
    /// written, instead of after it reached published data.
    #[test]
    fn the_summation_groups_are_disjoint() {
        groups_are_disjoint().expect("the ENDF summation rules partition these channels");
    }

    /// The disappearance set must not contain a channel that re-emits a
    /// neutron. Pinned by name, since these are the exact thirteen that were
    /// wrong before.
    #[test]
    fn disappearance_excludes_neutron_emitting_channels() {
        let abs = absorption();
        for mt in [11, 16, 17, 22, 23, 24, 25, 28, 29, 30, 32, 33, 34] {
            assert!(
                !abs.contains(&mt),
                "MT {mt} re-emits a neutron and cannot be a disappearance channel"
            );
        }
    }

    /// A huge absorption must not round away the scattering beside it.
    ///
    /// `scattering` in fast_xs.arrow was once reached as MT 2 + MT 3 - MT 27,
    /// which cancels the absorption back out through an intermediate the size
    /// of MT 3. These are TENDL-2025 Mo86's partials at 1.125e-5 eV: 2e12 barns
    /// of (n,p) and (n,alpha) against 37.8 barns of elastic, where that
    /// intermediate has an ulp of 4.9e-4 barns. The subtraction lands BELOW
    /// elastic, which a cross section that elastic is a part of cannot be.
    #[test]
    fn a_huge_absorption_does_not_round_away_the_scattering() {
        let elastic = 37.78729_f64;
        let partials: BTreeMap<i32, Vec<f64>> = [
            (2, vec![elastic]),
            (102, vec![1070.986]),
            (103, vec![2_053_910_000_000.0]),
            (107, vec![2_059_280_000_000.0]),
            (111, vec![416_142.1]),
        ]
        .into_iter()
        .collect();

        let derived = synthesize(&partials, 1);
        // No neutron-emitting channel but elastic is open this low, so the
        // scattering cross section IS the elastic one, exactly.
        let direct = elastic + non_elastic_scattering(&partials, 1)[0];
        assert_eq!(direct, elastic);

        // Pin the arithmetic that used to be here, so this cannot come back
        // looking like a harmless simplification.
        let by_subtraction = elastic + derived[&3][0] - derived[&27][0];
        assert!(
            by_subtraction < elastic,
            "MT 3 - MT 27 is supposed to lose precision here; if it no longer \
             does, this test is pinning nothing"
        );
    }

    #[test]
    fn a_threshold_reaction_is_zero_below_its_threshold() {
        let xs = on_grid(&[1.0, 2.0, 3.0], 2, 6);
        assert_eq!(xs, vec![0.0, 0.0, 1.0, 2.0, 3.0, 0.0]);
    }

    /// A cross section longer than the grid above its threshold is clipped
    /// rather than panicking, matching the Python writer.
    #[test]
    fn an_overlong_cross_section_is_clipped() {
        let xs = on_grid(&[1.0, 2.0, 3.0, 4.0], 2, 4);
        assert_eq!(xs, vec![0.0, 0.0, 1.0, 2.0]);
    }

    /// Total equals elastic plus non-elastic, and non-elastic equals the sum of
    /// everything else, on a hand-built nuclide where the answer is known.
    #[test]
    fn the_sums_add_up() {
        let n = 3;
        let mut partials = BTreeMap::new();
        partials.insert(2, vec![10.0; n]); // elastic
        partials.insert(51, vec![1.0; n]); // level inelastic
        partials.insert(16, vec![2.0; n]); // (n,2n), scattering
        partials.insert(102, vec![4.0; n]); // capture, disappearance
        let out = synthesize(&partials, n);

        assert_eq!(out[&4], vec![1.0; n], "MT 4 is the inelastic levels");
        assert_eq!(out[&101], vec![4.0; n], "MT 101 is the disappearance sum");
        assert_eq!(
            out[&27],
            vec![4.0; n],
            "MT 27 is fission plus disappearance"
        );
        assert_eq!(out[&3], vec![7.0; n], "MT 3 is 1 + 2 + 4");
        assert_eq!(out[&1], vec![17.0; n], "MT 1 is elastic plus non-elastic");
    }

    /// A redundant MT already present in the evaluation must not be summed back
    /// in. This is the double count that inflates every sum above it.
    #[test]
    fn an_evaluations_own_redundant_mts_are_not_summed_back_in() {
        let n = 2;
        let mut partials = BTreeMap::new();
        partials.insert(2, vec![10.0; n]);
        partials.insert(102, vec![4.0; n]);
        // The caller is required to exclude these; assert the sum is unchanged
        // if one slips through the door the other side.
        partials.insert(101, vec![999.0; n]);
        partials.insert(1, vec![999.0; n]);
        let out = synthesize(&partials, n);
        assert_eq!(out[&1], vec![14.0; n], "a stray MT 1 or 101 was summed in");
    }

    /// Fission comes from MT 18 where it exists, not from 18 and its partials.
    #[test]
    fn fission_is_not_counted_twice() {
        let n = 2;
        let mut partials = BTreeMap::new();
        partials.insert(18, vec![5.0; n]);
        partials.insert(19, vec![3.0; n]);
        partials.insert(20, vec![2.0; n]);
        let out = synthesize(&partials, n);
        assert_eq!(
            out[&27],
            vec![5.0; n],
            "MT 18 and its partials were both counted"
        );
    }
}

#[cfg(test)]
mod double_count_tests {
    use super::*;

    /// A total and its own residual levels must not both enter MT 3.
    ///
    /// The shape that found this: JEFF-4.0's Be9 has no evaluated MT 16, so the
    /// parser builds it as the sum of MT 875 to 890. Adding both the sum and
    /// its parts put the total cross section 35% above the evaluation's own
    /// MT 1, which cannot happen. Both converters did it, so a diff against the
    /// published data could not see it; only the physical bound could.
    #[test]
    fn a_total_and_its_own_levels_are_not_both_summed() {
        let n = 3;
        let mut partials: BTreeMap<i32, Vec<f64>> = BTreeMap::new();
        partials.insert(2, vec![10.0; n]); // elastic
                                           // (n,2n) given BOTH ways: the total and the two levels it sums.
        partials.insert(875, vec![1.0; n]);
        partials.insert(876, vec![2.0; n]);
        partials.insert(16, vec![3.0; n]);

        let out = synthesize(&partials, n);
        let mt3 = out.get(&3).expect("MT 3 is synthesized");
        assert_eq!(
            mt3[0], 3.0,
            "MT 3 must count (n,2n) once. Got {}, which is the 3 barns of \
             MT 875+876 plus another 3 from MT 16.",
            mt3[0]
        );

        // And the total must not exceed elastic plus one copy of the channel.
        let mt1 = out.get(&1).expect("MT 1 is synthesized");
        assert_eq!(mt1[0], 13.0, "MT 1 must be elastic plus MT 3, counted once");
    }

    /// With only the total present it is used, since there are no levels.
    #[test]
    fn a_total_with_no_levels_present_is_still_counted() {
        let n = 2;
        let mut partials: BTreeMap<i32, Vec<f64>> = BTreeMap::new();
        partials.insert(2, vec![10.0; n]);
        partials.insert(16, vec![3.0; n]);
        let out = synthesize(&partials, n);
        assert_eq!(
            out.get(&3).unwrap()[0],
            3.0,
            "MT 16 is the only form of (n,2n) here and must be counted"
        );
    }
}
